Rewrite type section

Girgias opened the pull request:
To incorporate typing features added in PHP 8.2.0 and also (hopefully) future proof for additional additions.

This is still a bit WIP as I haven't added DNF types but I already wanted to get some eyes on it to gather some opinions.
I've taken a bit of inspiration from the Rust manual: https://doc.rust-lang.org/reference/types.html

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -10,141 +10,131 @@
-    <literal>boolean</literal>, rather than of type <type>bool</type>:
+    The value of a scalar type declaration must be a value of the specified
+    type.
+    An exception exists that the <type>float</type> will always accept values
```
```suggestion
    An exception exists that the type <type>float</type> will always accept values
```

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,191 @@ Stack trace:
+  <sect3 xml:id="language.types.declarations.base.object">
+   <title><type>object</type> type</title>
+   <para>
+    The value must be an &instanceof; any class or enumeration case.
```
```suggestion
    The value must be an &instanceof; any class or enumeration.
```

Enumeration case would be the equivalent of an instance, so one of them is redundant.

Girgias replied:
Right, I wasn't totally sure about that so I've made the distinction, but good to know

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,191 @@ Stack trace:
+   </sect4>
+
+   <sect4 xml:id="language.types.declarations.base.class-specific.parent">
+    <title><type>self</type></title>
```
```suggestion
    <title><type>parent</type></title>
```

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,191 @@ Stack trace:
+    <title>static</title>
+    <para>
+     <type>static</type> is a return-only type which requires that the
+     value returned must be an &instanceof; the same class as the one
+     the method is called in.
```
```suggestion
     value returned must be an &instanceof; the same class as the class of the object itself.
```

Maybe the wording can be improved. But it has nothing to do with the caller, you can call said method outside of the given class hierarchy.

Girgias replied:
This is the current wording that already exist. I'll think about improving it

iluuu1994 replied:
Maybe we can explain with an example. "For example, an object `Cat` inheriting `Animal::getClone(): static` must return a `Cat`, not just any `Animal`, which would be the behavior of `self`."

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,191 @@ Stack trace:
+    <link linkend="language.types.declarations.composite.union">union type</link>
+    declaration. Available as of PHP 7.1.0.
+   </para>
+   <note>
```
Maybe a note could be added that `void` functions actually still return `null`.

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,191 @@ Stack trace:
+   <title>User-defined types</title>
+   <para>
+    <!-- todo links -->
+    A user-defined is any interface, class, or enumeration.
```
```suggestion
    A user-defined type is any interface, class, or enumeration.
```

iluuu1994 commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,191 @@ Stack trace:
+  </sect3>
+
+  <sect3 xml:id="language.types.declarations.base.function">
+   <title>Function types</title>
```
Maybe `\Closure` should be mentioned, and the fact that it is an object. It's also not completely clear we're talking about the `callable` type here.

Girgias replied:
We have a dedicated page on the callable type https://www.php.net/manual/en/language.types.callable.php and I thought about mentioning ``\Closure`` but couldn't totally figure out how to write this section.

iluuu1994 replied:
Ah ok, maybe just linking it should be fine then.

iluuu1994 reviewed (commented):
I'm not super familiar with the documentation format so some of these might be wrong (e.g. I'm not sure if `<type>` does something special).

cmb69 commented:
> I'm not sure if `<type>` does something special

Actually, that is supposed to create a link, but IIRC this is broken for quite a while now. Anyhow, could use [XML entities](https://github.com/php/doc-base/blob/ba9dbf211b61ea6e18fde8e01f52d9e8f4d8b59a/entities/global.ent#L556-L564) instead.

Crell commented on language/types/declarations.xml:
```diff
@@ -10,141 +10,131 @@
+     </itemizedlist>
+    </listitem>
+    <listitem>
+     <simpara>Function types</simpara>
```
We have function types?  I didn't think we had that yet.  I don't know what you mean otherwise.

Should `callable` also be in this list?

Girgias replied:
Well a ``callable`` is an extremely generic function type. And as I know @iluuu1994 is working on them I thought I could anticipate some doc writing. But I'm seeing this is being very confusing so I'll revert to using ``callable`` types for now.

Crell commented on language/types/declarations.xml:
```diff
@@ -10,141 +10,131 @@
+  </para>
 
-  <warning>
+  <para>
```
I'm not sure about moving this to a list instead of a table.  The table seems more readable to me for this, no?

Girgias replied:
I prefer the list as it allows nesting, also the table currently has no real grouping (and I don't know how to make it sensible) + a version release column which acts a bit weirdly live (might be some CSS issue)

Crell commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,197 @@ Stack trace:
+   <title>void</title>
+   <para>
+    <literal>void</literal> is a return-only type declaration indicating the
+    function does not return a value.
```
```suggestion
    function does not return a value, but the function may still terminate.
```

Crell commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,197 @@ Stack trace:
+   <title>Bottom type - never</title>
+   <para>
+    <literal>never</literal> is a return-only type indicating the function
+    does not return. This means that it either calls <function>exit</function>,
```
```suggestion
    does not terminate. This means that it either calls <function>exit</function>,
```

Crell commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,197 @@ Stack trace:
+   <note>
+    <simpara>
+     Even if a function has a return type of <literal>void</literal> it will
+     still return a value, this value is always &null;.
```
```suggestion
     still evaluate to a value in the calling code, specifically &null;.
```

iluuu1994 replied:
It actually does just return the value, just implicitly. So IMO this is accurate.

Crell replied:
Does it *return* a value, or *evaluate to* a value?  I've always understood it as "a function call that doesn't return evaluates to null".  Because a function with return void isn't allowed to have a return statement in it with any value at all, so it cannot return null, but it can evaluate to null by default.

(I don't know what the engine thinks of that, but that's how I've always conceptualized it from user space.)

Crell replied:
Huh.  TIL.  Also, PHP is screwy.

Crell commented on language/types/declarations.xml:
```diff
@@ -154,23 +144,197 @@ Stack trace:
+   <note>
+    <simpara>
+     It is not possible to define custom literal types.
+     Consider using an enumeration instead.
```
Link to the enum page?

Crell commented on language/types/declarations.xml:
```diff
@@ -10,141 +10,131 @@
-  <warning>
+  <para>
+   The list of base types is:
+   <itemizedlist>
```
Whether it's a list or stays as a table, each item should deep-link to the spot in the page where it's described.

Crell commented:
Glancing over the section... I barely understand why this page is what it is and where it is.  Shouldn't a description of different types come way earlier in the section?  It's basically at the end now.  Is the whole Types section due for an overhaul?

Girgias commented:
> Glancing over the section... I barely understand why this page is what it is and where it is. Shouldn't a description of different types come way earlier in the section? It's basically at the end now. Is the whole Types section due for an overhaul?

Yes, probably, I wrote this page back when PHP 8.0 got released as the information about type declarations where scattered in loads of different places. So this was done to centralise it.

Girgias commented:
> > I'm not sure if `<type>` does something special
> 
> Actually, that is supposed to create a link, but IIRC this is broken for quite a while now. Anyhow, could use [XML entities](https://github.com/php/doc-base/blob/ba9dbf211b61ea6e18fde8e01f52d9e8f4d8b59a/entities/global.ent#L556-L564) instead.

https://github.com/php/phd/pull/33 is meant to fix it, might need to go back to it

cmb69 reviewed (commented):
It seems to me we're mixing "types" and "type declarations", but we should better not. I think it's better to move much of what we have on the type declaration page to the [introduction page](https://www.php.net/manual/en/language.types.intro.php), and to have only info about type declarations on the former (general info about type declarations, and the special declaration types like `self`).

Girgias commented:
I don't really think I've addressed your comment @cmb69, but I'm still really blanking on how to break the relevant bits away, as I wrote it all from writing a type declaration perspective :-/

Girgias commented:
Okay, I've spent some time again rewritting this, and would love to get your comments and opinions on this.

cmb69 commented on language/types/type-juggling.xml:
```diff
@@ -169,18 +188,93 @@
+
+   <example>
+    <title>Example of types being coerced into a type part of the union</title>
+    <programlisting role="php">
+<![CDATA[
+<?php
```
This doesn't look like PHP code:
```suggestion
    <programlisting>
<![CDATA[
```
(accordingly, the `?>` needs to be removed as well)

cmb69 reviewed (commented):
Thank you! This is quite something.

Girgias commented on language/types.xml:
```diff
@@ -5,177 +5,134 @@
-   </listitem>
-
-  </itemizedlist>
+  <caution>
+   <simpara>
+    <link linkend="types.comparisons">The type comparison tables</link>
+    may also be useful, as various examples of comparison between values of
+    different types are present.
+   </simpara>
+  </caution>
```
Oh I forgot this existed, that is indeed better.

Girgias commented on language/types/type-system.xml:
```diff
@@ -0,0 +1,187 @@
+        <type>bool</type>, <type>int</type>, <type>float</type>, and <type>string</type>
+       </simpara>
+      </listitem>
+      <listitem>
+       <simpara><type>array</type> type</simpara>
+      </listitem>
+      <listitem>
+       <simpara><type>object</type> type</simpara>
+      </listitem>
```
For me, "primitive" is built-into the engine, and you cannot recreate them in userland. But I can change it.

Girgias commented:
Okay, I've only performed limited changes to try to address all the review comments. Hopefully this is now good to land and be stabbed by the pitchforks of translators.

@saundefined @mumumu what are your opinions of this as the translators of the 2 other main up to date languages?

saundefined commented:
> @saundefined @mumumu what are your opinions of this as the translators of the 2 other main up to date languages?

LGTM, thanks!

mumumu commented:
@Girgias Personally, I don't have a problem with it. This PR change is not mahoosive than I expected.

cmb69 reviewed (commented):
Thank you!

kitrio commented:
I found some type 
File  language/types/type-system.xml 
line 48 <simpara><type>float</type> string</simpara>
it seems like typo
because line 45 already write float   <simpara><type>float</type> type</simpara>



line 48 <simpara><type>string</type> type</simpara>
Is this right?

Thank you.

cmb69 commented:
Thanks for reporting, @kitrio! Should be fixed with https://github.com/php/doc-en/commit/23cda460978ca25b9c0be003862095cd0eb20b00.
